packages/peregrine/lib/List/useListState.js

```diff
@@ -1,179 +0,0 @@
-import { useMemo, useReducer, useCallback, useEffect } from 'react';
-
-import withLogger from '../util/withLogger';
-
-/**
- * Helper function to update selection of given key in the items list.
- *
- * @function updateSelection
- *
- * @param {Key} key
- * @param {Set} prevSelection
- * @param {Set} selectionModel
- * @returns {Set}
- */
-const updateSelection = (key, prevSelection, selectionModel) => {
-    let selection;
-    if (selectionModel === 'radio') {
-        selection = new Set().add(key);
-    }
-    if (selectionModel === 'checkbox') {
-        selection = new Set(prevSelection);
-        if (selection.has(key)) {
-            selection.delete(key);
-        } else {
-            selection.add(key);
-        }
-    }
-    return selection;
-};
-
-/**
- * Reducer function that is used by the useReducer hook inside the useListState hook.
- *
- * @function reducer
- *
- * @param {ListState} state
- * @param {Object} action
- * @param {string} action.type
- * @param {Object} action.payload
- * @param {Key} action.payload.key
- * @param {string} action.payload.selectionModel
- */
-const reducer = (state, { type, payload }) => {
-    switch (type) {
-        case 'REMOVE_FOCUS':
-            return {
-                ...state,
-                hasFocus: false
-            };
-        case 'SET_FOCUS':
-            return {
-                ...state,
-                hasFocus: true,
-                cursor: payload.key
-            };
-        case 'UPDATE_SELECTION': {
-            return {
-                ...state,
-                selection: updateSelection(
-                    payload.key,
-                    state.selection,
-                    payload.selectionModel
-                )
-            };
-        }
-        default:
-            return state;
-    }
-};
-
-const wrappedReducer = withLogger(reducer);
-
-const initialState = {
-    cursor: null,
-    hasFocus: false,
-    selection: new Set()
```

Comment:
The problem is `selection: new Set()`.

We'd have to pass in the `initialSelection` to the hook and it would have to either initialize `selection` using it, or add it later (before the first time `selection` is used).

Reply:
Yes, we can make `initialState` a function, something like `getInitialState` which takes the initial selection and uses that to create a new set.